UefiCpuPkg/Library/MpInitLib/DxeMpLib.c

```diff
@@ -565,17 +512,59 @@ RemoveNxprotection (
   EFI_STATUS                       Status;
   EFI_GCD_MEMORY_SPACE_DESCRIPTOR  MemDesc;
 
-  //
-  // TODO: Check EFI_MEMORY_XP bit set or not once it's available in DXE GCD
-  //       service.
-  //
   Status = gDS->GetMemorySpaceDescriptor (BaseAddress, &MemDesc);
   if (!EFI_ERROR (Status)) {
-    gDS->SetMemorySpaceAttributes (
-           BaseAddress,
-           Length,
-           MemDesc.Attributes & (~EFI_MEMORY_XP)
-           );
+    if ((MemDesc.Capabilities & EFI_MEMORY_XP) == EFI_MEMORY_XP ) {
+      if ((MemDesc.Attributes & EFI_MEMORY_XP) == EFI_MEMORY_XP) {
+        Status = gDS->SetMemorySpaceAttributes (
+                        BaseAddress,
+                        Length,
+                        MemDesc.Attributes & (~EFI_MEMORY_XP)
+                        );
+      }
+    }
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a - Removing Nx on 0x%p returned %r\n", __func__, BaseAddress, Status));
+    ASSERT_EFI_ERROR (Status);
+  }
+}
+
+/**
+  Add ReadOnly protection to the range specified by BaseAddress and Length.
+
+  The PEI implementation uses CpuPageTableLib to change the attribute.
+  The DXE implementation uses gDS to change the attribute.

```

Comment:
It is confusing to include the PEI implementation details here because this function is in `DxeMpLib.c`.

Reply:
cleaned up in latest.